src/humancellatlas/data/metadata/helpers/staging_area.py

```diff
@@ -0,0 +1,323 @@
+        metadata = {
+            'links.json': links_file.content
+        }
+        entity_ids_by_type = self.entity_ids_by_type(subgraph_id)
+        for entity_type, entity_ids in entity_ids_by_type.items():
+            # Sort entity_ids to produce the same ordering on multiple runs
+            for i, entity_id in enumerate(sorted(entity_ids)):
+                json_file_name = f'{entity_type}_{i}.json'
+                metadata_file = self.metadata[entity_id]
+                json_content = metadata_file.content
+                metadata[json_file_name] = json_content
+                file_manifest = {
+                    'content-type': 'application/json;',
+                    'crc32c': '0' * 8,
+                    'indexed': True,
+                    'name': json_file_name,
+                    's3_etag': None,
+                    'sha1': None,
+                    'sha256': '0' * 64,
+                    'size': len(json.dumps(json_content)),
+                    'uuid': metadata_file.uuid,
+                    'version': metadata_file.version
+                }
+                manifest.append(file_manifest)
+                if entity_type.endswith('_file'):
+                    file_manifest = self.descriptors[entity_id].to_manifest_entry
+                    manifest.append(file_manifest)
+                else:
+                    pass
+        logger.debug('Bundle contents: %s', list(metadata.keys()))
+        return links_file.version, manifest, metadata
+
+
+@attr.s(frozen=True, kw_only=True, auto_attribs=True)
+class GitHubStagingAreaFactory:
+    repo: Repository
+    branch: str
+    path: Tuple[str, ...]
+
+    @classmethod
+    def from_owner_and_name(cls,
+                            owner: str,
+                            name: str,
+                            branch: str,
+                            path: Optional[str] = None
+                            ) -> 'GitHubStagingAreaFactory':
+        """
+        :param owner: The owner of the GitHub repository
+        :param name: The name of the GitHub repository
+        :param branch: A branch name or commit SHA
+        :param path: The path inside the repository to the base of the staging area
+        """
+        token = os.environ['GH_API_TOKEN']  # A GitHub personal access token
+        logger.debug('Requesting GitHub repo %s', (owner, name, branch, path))
+        github_api = Github(token)
+        rate_limit = github_api.get_rate_limit()
+        logger.debug('GitHub API rate limit limit=%i, remaining=%i, reset=%s',
+                     rate_limit.core.limit, rate_limit.core.remaining, rate_limit.core.reset)
+        repo = github_api.get_repo(f'{owner}/{name}')
+        if len(branch) == 40 and set(branch).issubset(set('0123456789abcdef')):
+            branch_sha = branch
+        else:
+            gh_branch = repo.get_branch(branch)
+            branch_sha = gh_branch.commit.sha
+            logger.debug('Branch SHA %s', (branch, branch_sha))
```

Comment:
It's odd that you call the attribute `branch` but instead it only ever contains a sha, even though it can contain any ref so I don't understand why its called `branch` or why it is eagerly resolved, even though …

Reply:
Ah, I thought the branch value had to be converted to a corresponding commit SHA before it could be used to get the contents of a repository.